packages/hash/shared/src/blockMeta.ts

```diff
@@ -88,16 +86,41 @@ export const fetchBlockMeta = async (
   }
 
   const promise = (async () => {
-    const metadata: BlockMetadata = await (
-      await fetch(`${url}/block-metadata.json`)
-    ).json();
-
-    const schema: Block["componentSchema"] = await (
-      await fetch(`${url}/${metadata.schema}`)
-    ).json();
+    // the spec requires a metadata file called `block-metadata.json`
+    const metadataUrl = `${url}/block-metadata.json`;
+    let metadata: BlockMetadata;
+    try {
+      metadata = await (await fetch(metadataUrl)).json();
+    } catch (err) {
+      blockCache.delete(url);
+      throw new Error(
+        `Could not fetch and parse block metadata at url ${metadataUrl}: ${
+          (err as Error).message
+        }`,
+      );
+    }
+
+    const schemaPath = metadata.schema;
+
+    // schema urls may be absolute, as blocks may rely on schemas they do not define
+    const schemaUrl =
+      schemaPath && schemaPath.match(/^(?:[a-z]+:)?\/\//)
+        ? schemaPath
+        : `${url}/${schemaPath.replace(/^\//, "")}`;
```

Comment:
I can imagine unhandled errors in these lines too, e.g. if `metadata.schema` is not defined. `await (await fetch(metadataUrl)).json();` returns `any`, which is why TypeScript does not complain. We can use `json-schema` to reject invalid JSONs early on. Either way, schema URL extraction can be moved into `try / catch`.

We can use `URL` API to avoid hand-written conditions for relative links. Examples:
https://developer.mozilla.org/en-US/docs/Web/API/URL#usage_notes

In our case it’d be something like:

```ts
const schemaUrl = new URL(metadata.schema.replace(/^\//, ""), url);
```

Reply:
Ah, very handy! I did not know that. Added in https://github.com/hashintel/hash/pull/449/commits/3d3edb0bb8ae73f80b698779c3d810cc74f57ad1

I'll leave 'is this a valid JSON schema' for another time

Reply:
Ah, very handy! I did not know that. Added in https://github.com/hashintel/hash/pull/449/commits/3d3edb0bb8ae73f80b698779c3d810cc74f57ad1

I'll leave 'is this a valid JSON schema' for another time

Reply:
I've undone the change to use the `URL` constructor because it only takes the origin of the provided base - our base URLs are full paths (e.g. "https://blockprotocol.org/blocks/@hash/table"), and to use the constructor we'd have to split off the rest of the path and add it to the schema path. This is about as much manual manipulation as was there before, without creating new objects, so I've just reverted it to how it was (but inside the `try` block). let me know if I'm missing a trick here

Reply:
Ah that’s a shame. I first thought it was to do with a trailing slash, but Node REPL proved me wrong:

```js
new URL("/bar", "https://example.com/foo").toString() // https://example.com/bar
new URL("bar",  "https://example.com/foo").toString() // https://example.com/bar
```

🙁

Reply:
Yep, I went through the same process :grin: